### ISSUE TITLE
Wheels broken on OS-X 

### ISSUE THREAD

**Issue by ChrisBarker-NOAA** [kyamagu/skia-python#137]
When I install via pip, I get the wheels published on PyPi. Example:

skia_python-87.1-cp39-cp39-macosx_10_14_x86_64.whl

However, they don't work :-( Erroring out on import with:

```
ImportError: dlopen(/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/skia.cpython-39-darwin.so, 2): Symbol not found: _XML_ErrorString
  Referenced from: /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/skia.cpython-39-darwin.so
  Expected in: flat namespace
 in /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/skia.cpython-39-darwin.so
```

Tested under the Python.org 3.9 installer, and conda-forge 3.8 and 3.9 -- same results on all three.

It seems that it's looking for the _XML_ErrorString symbol from expat, which is bundled with Python:

```
$ python -c "import xml.parsers.expat; print xml.parsers.expat.__version__"
$Revision: 17640 $
```

I'm totally guessing but I think it might be the "flat namespace" issue. The symbol is there in the extension:

```
$ grep "_XML_ErrorString" /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/lib-dynload/pyexpat.cpython-39-darwin.so 
Binary file /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/lib-dynload/pyexpat.cpython-39-darwin.so matches
```

Sorry -- I've reached the end of what I know about these issues.

**Comment by kyamagu** [comment]
Thanks for reporting this, is this related to xcode issue?
https://github.com/pyenv/pyenv/issues/544

**Comment by ChrisBarker-NOAA** [comment]
related, maybe, but I don't think it's the same.

I just took a look at the setup.py, and can't find an issue. I also looked at what the extension is linked to:

otool -L skia.cpython-39-darwin.so 
skia.cpython-39-darwin.so:
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.100.1)
	/System/Library/Frameworks/AppKit.framework/Versions/C/AppKit (compatibility version 45.0.0, current version 1894.60.100)
	/System/Library/Frameworks/ApplicationServices.framework/Versions/A/ApplicationServices (compatibility version 1.0.0, current version 52.0.0)
	/System/Library/Frameworks/OpenGL.framework/Versions/A/OpenGL (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 902.1.0)
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1677.104.0)
	/System/Library/Frameworks/CoreGraphics.framework/Versions/A/CoreGraphics (compatibility version 64.0.0, current version 1355.22.0)
	/System/Library/Frameworks/CoreServices.framework/Versions/A/CoreServices (compatibility version 1.0.0, current version 1069.24.0)
	/System/Library/Frameworks/CoreText.framework/Versions/A/CoreText (compatibility version 1.0.0, current version 1.0.0)

And I don't see anything that looks odd.

But apparently, it's looking for expat somewhere that it's not finding it, or it's finding a different version.

How are the wheels built? i.e. what Python version, what OS-X version? IN your tox setup you have tests, so presumably it's working somewhere.

Are you linking in expat anywhere explicitly? I can't find that in the setup.py

**Comment by kyamagu** [comment]
See CI config in .github/workflows/--
山口光太
（株）サイバーエージェント AI Lab
050-5305-9404
yamaguchi_kota@cyberagent.co.jp

**Comment by ChrisBarker-NOAA** [comment]
OK -- so I see that it's using "macos-latest", I wonder what that is?

Then I see:
-       matrix:
        python-version: [3.7]

but how do the other wheels get built ?

Anyway, I see no reference to expat in the skia build command. and in here:

https://github.com/google/skia/blob/master/third_party/expat/BUILD.gn

I see: ```
declare_args() {
  skia_use_system_expat = is_official_build
}
```
and in the CI skia build line, I see: `is_official_build=true` -- so I think that means that it's using the "system" expat.

I theory, expat has shipped with OS-X for quite some time, so it should be there, and setting teh deployment target is supposed to work, but apparently not in this case :-(

Maybe setting skia_use_system_expat to false will bundle it and solve this issue ?

**Comment by ChrisBarker-NOAA** [comment]
One more noteL I tried the wheels with python3.7 -- same error.

**Comment by kyamagu** [comment]
Bundling expat might work, but I suspect something is not working in the system side in your environment, like Xcode.
CI testing works normal https://github.com/kyamagu/skia-python/actions/workflows/build.yml

According to Github Actions, macos-latest runner uses macOS Catalina 10.15 with 3-core CPU https://docs.github.com/en/actions/reference/specifications-for-github-hosted-runners

**Comment by ChrisBarker-NOAA** [comment]
Well, I have yet to get it to build locally -- is there a script for the whole thing, or do I need to replicate what I see in the CI?

Anyway, I'm running on 10.14 right now, and if it was built with 10.15 that might be the problem. However, the wheels are:

skia_python-87.1-cp38-cp38-macosx_10_14_x86_64.whl 

which means they should run on any 10.14 or greater mac -- XCode should not need to be installed. So there's something not quite right with the wheel building.

**Comment by kyamagu** [comment]
CI description are the exact steps to reproduce https://github.com/kyamagu/skia-python/blob/master/.github/workflows/build.yml

You might want to try a local runner tool https://github.com/nektos/act

In any case, build requires `clang` which does require Xcode command line tools.

**Comment by ChrisBarker-NOAA** [comment]
The CI is building and testing on 10.5, so maybe the wheels work there, but not on 10.14 :-(

I do see a lot of these warnings:

warning: object file (skia/out/Release/libskia.a(libjpeg.jutils.o)) was built for newer macOS version (10.15) than being linked (10.9)

So perhaps that's the source of the problem.

I was wanting to give this a quick try to see if it was worth pursing -- so I'm not sure if I'll be able to get around to debugging these build issues -- but good luck!

**Comment by justvanrossum** [comment]
I can confirm that the skia-python wheels provided by PyPI currently do not work on 10.14, with the error as reported by Chris: https://github.com/justvanrossum/drawbot-skia/issues/37

**Comment by kyamagu** [comment]
v87.3 might have fixed this issue, though I cannot test

**Comment by kyamagu** [comment]
ok, thanks--
山口光太
（株）サイバーエージェント AI Lab
050-5305-9404
***@***.***

**Comment by kyamagu** [comment]
You can check if the statically linked libexpat solves this issue. Download the artifact of this build:
https://github.com/kyamagu/skia-python/actions/runs/1209672577

**Comment by justvanrossum** [comment]
> You can check if the statically linked libexpat solves this issue

It does!

**Comment by kyamagu** [comment]
Fixed in #152
